firewood/src/merkle/changes.rs

```diff
@@ -0,0 +1,389 @@
+                    let path_overlap = PrefixOverlap::from(key, child_key.as_ref());
+                    let unique_node = path_overlap.unique_b;
+                    if unique_node.is_empty() || child_key > *key {
+                        self.traversal_stack.push(NodeState::new(
+                            child_pre_path,
+                            child,
+                            self.trie,
+                        )?);
+                        // Once we have found the first child that should be traversed, we can break out of
+                        // the loop where we we add the rest of the children without the above test.
+                        break;
+                    }
+                }
+
+                // Add the rest of the children (if any) to the traversal stack.
+                for node_state in reversed_children_with_pre_path
+                    .map(|(child, child_pre_path)| NodeState::new(child_pre_path, child, self.trie))
+                {
+                    self.traversal_stack.push(node_state?);
+                }
+            }
+
+            // Pop the next node state from the stack
+            if let Some(state) = self.traversal_stack.pop() {
+                // Since pre-order traversal of a trie iterates through the nodes in lexicographical
+                // order, we can stop the traversal once we see a node key that is larger than or
+                // equal to the key. We stop the traversal by pushing the current `NodeState`
+                // back to the stack. Calling `next` will process this node.
+                let node_key = key_from_nibble_iter(state.path.iter().copied());
+                if node_key >= *key {
+                    self.traversal_stack.push(state);
+                    return Ok(());
+                }
+
+                // If this node is a leaf, then we don't need to save its state to `node_state` as
+                // it has no children to traverse.
+                if let Node::Branch(_branch) = &*state.node {
+                    // Check if this node's path is a prefix of the key. If it is not (`unique_node`
+                    // is not empty), then this node's children cannot be larger than or equal to
+                    // the key, and we don't need to include them on the traversal stack.
+                    let path_overlap = PrefixOverlap::from(key, node_key.as_ref());
+                    let unique_node = path_overlap.unique_b;
+                    if unique_node.is_empty() {
+                        self.node_state = Some(state);
+                    }
+                }
+            } else {
+                // Traversal stack is empty. This means the key is lexicographically larger than
+                // all of the keys in the trie. Calling `next` will return None.
+                return Ok(());
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(
```

Comment:
nit: expect, so these don't go stale. Not so important in tests, but still a good habit.